File: src/main/java/ru/vood/freemarker/ext/sql/FetchedResultSet.java
package ru.vood.freemarker.ext.sql;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;
import java.util.stream.Collectors;

public class FetchedResultSet {
    private final List<ColumnMeta> meta;
    private final Map<String, Integer> columnIndices;
    private final List<List<Object>> rows;

    FetchedResultSet(ResultSet rs, FtlDefaultObjectWrapper ftlDefaultObjectWrapper) throws SQLException {
        try {
            this.meta = Collections.unmodifiableList(ColumnMeta.of(rs.getMetaData()));
            final int columnCount = meta.size();
            Map<String, Integer> tempColumnIndices = new HashMap<>(columnCount, 1);
            for (int i = 0; i < columnCount; i++) {
                String columnLabel = meta.get(i).getColumnLabel();
                // if two columns have same labels, save only the 1st one
                tempColumnIndices.putIfAbsent(columnLabel, i);
            }
            List<List<Object>> tempRows = new ArrayList<>(64);
            while (rs.next()) {
                List<Object> row = new ArrayList<>(columnCount);
                for (int i = 0; i < columnCount; i++) {
                    Object o = rs.getObject(i + 1);
                    if (o instanceof ResultSet) {
                        o = new FetchedResultSet((ResultSet) o, ftlDefaultObjectWrapper);
                    }
                    row.add(o);
                }
                tempRows.add(Collections.unmodifiableList(row));
            }
            this.rows = Collections.unmodifiableList(tempRows);
            this.columnIndices = Collections.unmodifiableMap(tempColumnIndices);
        } finally {
            if (!rs.isClosed()) {
                rs.close();
            }
        }
    }

    public List<String> getColumnLabels() {
        return meta.stream().map(ColumnMeta::getColumnLabel).collect(Collectors.toList());
    }

    public Integer getColumnIndex(String label) {
        return columnIndices.get(label);
    }

    public Integer findColumn(String label) {
        Integer index = getColumnIndex(label);
        if (index == null) return null;
        return index + 1;
    }

    public List<ColumnMeta> getMetaData() {
        return meta;
    }

    public Map<String, Integer> getColumnIndices() {
        return columnIndices;
    }

    public List<List<Object>> getRows() {
        return rows;
    }
}

File: src/main/java/ru/vood/freemarker/ext/sql/FetchedResultSet.kt
package ru.vood.freemarker.ext.sql

import ru.vood.freemarker.ext.sql.ColumnMeta
import java.sql.ResultSet
import java.util.*
import java.util.stream.Collectors

/**
 * This class contains the static result of fetching from a [ResultSet] and its metadata. This is needed because
 * the original result set can be fetched only once, so wrapping it directly may lead to getting empty result set in
 * FTL if the object is accessed twice.
 */
class FetchedResultSet internal constructor(rs: ResultSet, ftlDefaultObjectWrapper: FtlDefaultObjectWrapper) {
    var metaData: List<ColumnMeta>
    var columnIndices: Map<String, Int>
    var rows: List<List<Any>>
    val columnLabels: List<String>
        get() = metaData.stream().map(ColumnMeta::columnLabel).collect(Collectors.toList())

    fun getColumnIndex(label: String): Int {
        return columnIndices[label]!!
    }

    fun findColumn(label: String): Int? {
        val index = getColumnIndex(label)
        return index + 1
    }

    init {
        try {
            metaData = Collections.unmodifiableList(ColumnMeta.of(rs.metaData))
            val columnCount = metaData.size
            val tempColumnIndices: MutableMap<String, Int> = HashMap(columnCount)
            for (i in 0 until columnCount) {
                val columnLabel = metaData[i].columnLabel
                // if two columns have same labels, save only the 1st one
                tempColumnIndices.putIfAbsent(columnLabel, i)
            }
            val tempRows: MutableList<List<Any>> = ArrayList(64)
            while (rs.next()) {
                val row: MutableList<Any> = ArrayList(columnCount)
                for (i in 0 until columnCount) {
                    var o = rs.getObject(i + 1)
                    if (o is ResultSet) {
                        o = FetchedResultSet(o, ftlDefaultObjectWrapper)
                    }
                    row.add(o)
                }
                tempRows.add(Collections.unmodifiableList(row))
            }
            rows = Collections.unmodifiableList(tempRows)
            columnIndices = Collections.unmodifiableMap(tempColumnIndices)
        } finally {
            if (!rs.isClosed) {
                rs.close()
            }
        }
    }
}